numpy/lib/function_base.py

```diff
@@ -3929,21 +3928,35 @@ def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
             indices_above = indices_above[:-1]
             n = np.isnan(ap[-1:, ...])
 
-        x1 = take(ap, indices_below, axis=axis) * weights_below
-        x2 = take(ap, indices_above, axis=axis) * weights_above
+        x1 = take(ap, indices_below, axis=axis)
+        x2 = take(ap, indices_above, axis=axis)
+
+        # the linear interpolation below ensure monotonicity with floating
+        # point operations and was proposed in:
+        # https://math.stackexchange.com/a/1798323 by Pedro Gimeno
+        diff_x2_x1 = x2 - x1
+        r_above = x1 + diff_x2_x1 * weights_above
+        r_below = x2 - diff_x2_x1 * weights_below
 
         # ensure axis with q-th is first
-        x1 = np.moveaxis(x1, axis, 0)
-        x2 = np.moveaxis(x2, axis, 0)
```

Comment:
If you kept these lines, and did them before the computation, then you'd be able to do:
```
r_above = np.add(x1, diff_x2_x1 * weights_above, out=out)
r_below = np.subtract(x2, diff_x2_x1 * weights_below, out=r_above, where=weights_above < 0.5)
```